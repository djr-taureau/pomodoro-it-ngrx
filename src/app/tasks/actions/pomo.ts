import { Action } from '@ngrx/store';
import { Pomo } from '../models/pomo';

export enum PomoActionTypes {
  Load = '[Pomo] Load',
  LoadSuccess = '[Pomo] Load Success',
  LoadFail = '[Pomo] Load Fail',
  Select = '[Pomo] Select',
}

export class Load implements Action {
  readonly type = PomoActionTypes.Load;
  constructor(public payload: Pomo) {}
}
export class LoadSuccess implements Action {
  readonly type = PomoActionTypes.LoadSuccess;
  constructor(public payload: Pomo[]) {}
}

export class LoadFail implements Action {
  readonly type = PomoActionTypes.LoadSuccess;
  constructor(public payload: any) {}
}
export class Select implements Action {
  readonly type = PomoActionTypes.Select;

  constructor(public payload: string) {}
}

export type PomoActions =
  | Load
  | LoadSuccess
  | LoadFail
  | Select;
