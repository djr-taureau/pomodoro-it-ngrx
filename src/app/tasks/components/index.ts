import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { ReactiveFormsModule } from '@angular/forms';
import { RouterModule } from '@angular/router';
import { TaskDetailComponent } from './task-detail';
import { TaskPreviewComponent } from './task-preview';
import { TaskPreviewListComponent } from './task-preview-list';
import { TaskSearchComponent } from './task-search';
import { PomoTrackerComponent } from './pomo-tracker.component';
import { PipesModule } from '../../shared/pipes';
import { MaterialModule } from '../../material';

export const COMPONENTS = [
  TaskDetailComponent,
  TaskPreviewComponent,
  TaskPreviewListComponent,
  TaskSearchComponent,
  PomoTrackerComponent
];

@NgModule({
  imports: [
    CommonModule,
    ReactiveFormsModule,
    MaterialModule,
    RouterModule,
    PipesModule,
  ],
  declarations: COMPONENTS,
  exports: COMPONENTS,
})
export class ComponentsModule {}
