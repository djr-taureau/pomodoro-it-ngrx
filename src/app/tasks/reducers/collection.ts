import {
  CollectionActionTypes,
  CollectionActions,
} from './../actions/collection';

export interface State {
  loaded: boolean;
  loading: boolean;
  ids: string[];
  pomoIds: string[];
}

const initialState: State = {
  loaded: false,
  loading: false,
  ids: [],
  pomoIds: []
};

export function reducer(
  state = initialState,
  action: CollectionActions
): State {
  switch (action.type) {

    case CollectionActionTypes.Load: {
      return {
        ...state,
        loading: true,
      };
    }



    case CollectionActionTypes.LoadSuccess: {
      return {
        loaded: true,
        loading: false,
        ids: action.payload.map(task => task.id),
        pomoIds: []
      };
    }

    case CollectionActionTypes.LoadPomosSuccess: {
      return {
        loaded: true,
        loading: false,
        pomoIds: action.payload.map(pomo => pomo.id),
        ids: []
      };
    }

    // case CollectionActionTypes.LoadPomosSuccess: {
    //   return {
    //     loaded: true,
    //     loading: false,
    //     pomoIds: action.payload.map(pomo => pomo.id),
    //     ids: []
    //   };
    // }

    case CollectionActionTypes.AddTaskSuccess:
    case CollectionActionTypes.RemoveTaskFail: {
      if (state.ids.indexOf(action.payload.id) > -1) {
        return state;
      }

      return {
        ...state,
        ids: [...state.ids, action.payload.id],
      };
    }

    case CollectionActionTypes.RemoveTaskSuccess:
    case CollectionActionTypes.AddTaskFail: {
      return {
        ...state,
        ids: state.ids.filter(id => id !== action.payload.id),
      };
    }

    default: {
      return state;
    }
  }
}


export const getLoaded = (state: State) => state.loaded;

export const getLoading = (state: State) => state.loading;

export const getIds = (state: State) => state.ids;

export const getPomoIds = (state: State) => state.pomoIds;

