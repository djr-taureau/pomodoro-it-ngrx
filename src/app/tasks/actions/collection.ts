import { Action } from '@ngrx/store';
import { Task } from '../models/task';
import { Pomo } from '../models/pomo';

export enum CollectionActionTypes {
  AddTask = '[Collection] Add Task',
  AddTaskSuccess = '[Collection] Add Task Success',
  AddTaskFail = '[Collection] Add Task Fail',
  RemoveTask = '[Collection] Remove Task',
  RemoveTaskSuccess = '[Collection] Remove Task Success',
  RemoveTaskFail = '[Collection] Remove Task Fail',
  Load = '[Collection] Load',
  LoadSuccess = '[Collection] Load Success',
  LoadFail = '[Collection] Load Fail',
  LoadPomos = '[Pomos] Load',
  LoadPomosSuccess = '[Pomos] Load Success',
  LoadPomosFail = '[Pomos] Load Fail',
}

/**
 * Add Task to Collection Actions
 */
export class AddTask implements Action {
  readonly type = CollectionActionTypes.AddTask;

  constructor(public payload: Task) {}
}

export class AddTaskSuccess implements Action {
  readonly type = CollectionActionTypes.AddTaskSuccess;

  constructor(public payload: Task) {}
}

export class AddTaskFail implements Action {
  readonly type = CollectionActionTypes.AddTaskFail;

  constructor(public payload: Task) {}
}

/**
 * Remove Task from Collection Actions
 */
export class RemoveTask implements Action {
  readonly type = CollectionActionTypes.RemoveTask;

  constructor(public payload: Task) {}
}

export class RemoveTaskSuccess implements Action {
  readonly type = CollectionActionTypes.RemoveTaskSuccess;

  constructor(public payload: Task) {}
}

export class RemoveTaskFail implements Action {
  readonly type = CollectionActionTypes.RemoveTaskFail;

  constructor(public payload: Task) {}
}

/**
 * Load Collection Actions
 */
export class Load implements Action {
  readonly type = CollectionActionTypes.Load;
}


export class LoadSuccess implements Action {
  readonly type = CollectionActionTypes.LoadSuccess;

  constructor(public payload: Task[]) {}
}

export class LoadFail implements Action {
  readonly type = CollectionActionTypes.LoadFail;

  constructor(public payload: any) {}
}

export class LoadPomos implements Action {
  readonly type = CollectionActionTypes.LoadPomos;
}

export class LoadPomosSuccess implements Action {
  readonly type = CollectionActionTypes.LoadPomosSuccess;

  constructor(public payload: Pomo[]) {}
}

export class LoadPomosFail implements Action {
  readonly type = CollectionActionTypes.LoadFail;

  constructor(public payload: any) {}
}

export type CollectionActions =
  | AddTask
  | AddTaskSuccess
  | AddTaskFail
  | RemoveTask
  | RemoveTaskSuccess
  | RemoveTaskFail
  | Load
  | LoadSuccess
  | LoadFail
  | LoadPomos
  | LoadPomosSuccess
  | LoadPomosFail;

