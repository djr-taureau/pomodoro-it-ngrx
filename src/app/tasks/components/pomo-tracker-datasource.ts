import { getPomosEntitiesState } from './../reducers/index';
import { DataSource } from '@angular/cdk/collections';
import { MatPaginator, MatSort } from '@angular/material';
import { map, concatMap, flatMap, reduce, scan, subscribeOn } from 'rxjs/operators';
import { BehaviorSubject, Observable, of as observableOf, merge } from 'rxjs';
import { Pomo } from '../../tasks/models/pomo';
export class PomoTrackerDataSource extends DataSource<Pomo> {
  sort;
  paginator;

  dataStream: BehaviorSubject<Pomo[]> = new BehaviorSubject<Pomo[]>([]);
  set data(v: Pomo[]) { this.dataStream.next(v); }
  get data(): Pomo[] { return this.dataStream.value; }
  // data() {
  //   return this.dataStream.value;
  // }

  constructor(data: Pomo[], sort: MatSort, paginator: MatPaginator) {
    super();
    this.dataStream.next(data);
    // this.data = this.dataStream.next(data);
  }
  /**
   * Connect this data source to the table. The table will only update when
   * the returned stream emits new items.
   * @returns A stream of the items to be rendered.
   * @memberof PomoTrackerDataSource
  /*
   */
  connect(): Observable<Pomo[]> {

    const dataMutations = [
      observableOf(this.data),
      this.paginator.page,
      this.sort.sortChange
    ];

    // Set the paginators length
    this.paginator.length = this.data.length;

    return merge(...dataMutations).pipe(map(() => {
      return this.getPagedData(this.getSortedData([...this.data]));
    }));
    // return this.dataStream.asObservable();
  }

  disconnect() {
    this.dataStream.complete();
  }

  /**
   * Paginate the data (client-side). If you're using server-side pagination,
   * this would be replaced by requesting the appropriate data from the server.
   */
  private getPagedData(data: Pomo[]) {
    const startIndex = this.paginator.pageIndex * this.paginator.pageSize;
    return data.splice(startIndex, this.paginator.pageSize);
  }

  /**
   * Sort the data (client-side). If you're using server-side sorting,
   * this would be replaced by requesting the appropriate data from the server.
   */
  private getSortedData(data: Pomo[]) {
    if (!this.sort.active || this.sort.direction === '') {
      return data;
    }

    return data.sort((a, b) => {
      const isAsc = this.sort.direction === 'asc';
      switch (this.sort.active) {
        case 'date': return compare(a.date, b.date, isAsc);
        case 'notes': return compare(+a.notes, +b.notes, isAsc);
        default: return 0;
      }
    });
  }

}

/** Simple sort comparator for example ID/Name columns (for client-side sorting). */
function compare(a, b, isAsc) {
  return (a < b ? -1 : 1) * (isAsc ? 1 : -1);
}
