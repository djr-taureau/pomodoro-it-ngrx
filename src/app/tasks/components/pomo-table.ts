import { Component, OnInit, ViewChild, Input, AfterViewInit, EventEmitter, ChangeDetectionStrategy } from '@angular/core';
import { MatPaginator, MatSort, MatTableDataSource } from '@angular/material';
import { Pomo } from '../models/pomo';
import { from as observableFrom, of as observableOf, Observable } from 'rxjs';
import { mergeScan, scan, concatMap, mergeMap, flatMap, map } from 'rxjs/operators';
import { Store, select } from '@ngrx/store';
import * as fromTasks from '../reducers';
import * as fromPomos from '../reducers';
import * as collection from '../actions/collection';

@Component({
  selector: 'app-pomo-table',
  template: `
  <div class="mat-elevation-z8">
  <mat-table #table [dataSource]="this.pomos$ | async" matSort aria-label="Pomos">

    <!-- Date Column -->
    <ng-container matColumnDef="date">
      <mat-header-cell *matHeaderCellDef mat-sort-header>Date</mat-header-cell>
      <mat-cell *matCellDef="let row">{{row.date | date: 'short'}}</mat-cell>
    </ng-container>

    <!-- Notes Column -->
    <ng-container matColumnDef="notes">
      <mat-header-cell *matHeaderCellDef mat-sort-header>Notes</mat-header-cell>
      <mat-cell *matCellDef="let row">{{row.notes}}</mat-cell>
    </ng-container>

    <mat-header-row *matHeaderRowDef="displayedColumns"></mat-header-row>
    <mat-row *matRowDef="let row; columns: displayedColumns;"></mat-row>
  </mat-table>

  <mat-paginator #paginator
    [length]="dataSource.data.length"
    [pageIndex]="0"
    [pageSize]="50"
    [pageSizeOptions]="[25, 50, 100, 250]">
  </mat-paginator>
</div>
  `,
  styles: [
    `
    :host {
      display: flex;
    }

    :host a {
      display: flex;
    }

    mat-card {
      width: 400px;
      margin: 15px;
      display: flex;
      flex-flow: column;
      justify-content: space-between;
    }

    @media only screen and (max-width: 768px) {
      mat-card {
        margin: 15px 0 !important;
      }
    }
    mat-card:hover {
      box-shadow: 3px 3px 16px -2px rgba(0, 0, 0, .5);
    }
    mat-card-title {
      margin-right: 10px;
    }
    mat-card-title-group {
      margin: 0;
    }
    a {
      color: inherit;
      text-decoration: none;
    }
    img {
      width: 60px;
      min-width: 60px;
      margin-left: 5px;
    }
    mat-card-content {
      margin-top: 15px;
      margin: 15px 0 0;
    }
    span {
      display: inline-block;
      font-size: 13px;
    }
    mat-card-footer {
      padding: 0 25px 25px;
    }
  `,
  ],
})
export class PomoTableComponent implements OnInit {
  @Input() pomos$: Observable<Pomo[]>;
  @ViewChild(MatSort) sort: MatSort;
  @ViewChild(MatPaginator) paginator: MatPaginator;
  renderedData: Pomo[];
  displayedColumns = ['date', 'notes'];

ngOnInit() {
  //
}
}
