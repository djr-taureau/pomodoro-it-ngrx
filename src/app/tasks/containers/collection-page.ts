import { Component, ChangeDetectionStrategy, OnInit } from '@angular/core';
import { Store, select } from '@ngrx/store';
import { Observable } from 'rxjs/Observable';

import * as fromTasks from '../reducers';
import * as pomos from '../actions/pomo';
import * as collection from '../actions/collection';
import { Task } from '../models/task';

@Component({
  selector: 'app-collection-page',
  changeDetection: ChangeDetectionStrategy.OnPush,
  template: `
    <mat-card>
      <mat-card-title>My Pomodo-it Tasks</mat-card-title>
    </mat-card>

    <app-task-preview-list [tasks]="tasks$ | async"></app-task-preview-list>
  `,
  styles: [
    `
    mat-card-title {
      display: flex;
      justify-content: center;
    }
  `,
  ],
})
export class CollectionPageComponent implements OnInit {
  tasks$: Observable<Task[]>;

  constructor(private store: Store<fromTasks.State>) {
    this.tasks$ = store.pipe(select(fromTasks.getTaskCollection));
  }

  ngOnInit() {
    this.store.dispatch(new collection.Load());
    this.store.dispatch(new collection.LoadPomos());
  }
}
