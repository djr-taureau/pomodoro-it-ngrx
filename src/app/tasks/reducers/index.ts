import {
  createSelector,
  createFeatureSelector,
  ActionReducerMap,
  MemoizedProjection,
  MemoizedSelector
} from '@ngrx/store';
import * as fromSearch from './search';
import * as fromTasks from './tasks';
import * as fromPomos from './pomos';
import * as fromCollection from './collection';
import * as fromRoot from '../../reducers';
import { createEntityAdapter } from '@ngrx/entity';
import { getSelectedId } from './tasks';
import { StoreRouterConnectingModule, RouterStateSerializer, routerReducer } from '@ngrx/router-store';
import { Params } from '@angular/router';
import { Pomo } from '../models/pomo';
import { Task } from '../models/task';
export interface TasksState {
  search: fromSearch.State;
  tasks: fromTasks.State;
  pomos: fromPomos.State;
  collection: fromCollection.State;
}

export interface State extends fromRoot.State {
  tasks: TasksState;
}

export const reducers: ActionReducerMap<TasksState> = {
  search: fromSearch.reducer,
  tasks: fromTasks.reducer,
  pomos: fromPomos.reducer,
  collection: fromCollection.reducer,
};

/**
 * A selector function is a map function factory. We pass it parameters and it
 * returns a function that maps from the larger state tree into a smaller
 * piece of state. This selector simply selects the `Tasks` state.
 *
 * Selectors are used with the `select` operator.
 *
 * ```ts
 * class MyComponent {
 *   constructor(state$: Observable<State>) {
 *     this.TasksState$ = state$.pipe(select(getTasksState));
 *   }
 * }
 * ```
 */

/**
 * The createFeatureSelector function selects a piece of state from the root of the state object.
 * This is used for selecting feature states that are loaded eagerly or lazily.
 */
export const getTasksState = createFeatureSelector<TasksState>('tasks');

export const getTaskEntitiesState = createSelector(
  getTasksState,
  state => state.tasks
);

export const {
  selectIds: getTaskIds,
  selectEntities: getTaskEntities,
  selectAll: getAllTasks,
  selectTotal: getTotalTasks,
} = fromTasks.adapter.getSelectors(getTaskEntitiesState);

export const getPomosEntitiesState = createSelector(
  getTasksState,
  state => state.pomos
);

export const {
  selectIds: getPomoIds,
  selectEntities: getPomoEntities,
  selectAll: getAllPomos,
  selectTotal: getTotalPomos,
} = fromPomos.adapter.getSelectors(getPomosEntitiesState);


export const getSelectedTaskId = createSelector(
  getTaskEntitiesState,
  fromTasks.getSelectedId
);

export const getSelectedTask = createSelector(
  getTaskEntities,
  getSelectedTaskId,
  (entities, selectedId) => {
    return selectedId && entities[selectedId];
  }
);

export const getSearchState = createSelector(
  getTasksState,
  (state: TasksState) => state.search
);

export const getSearchTaskIds = createSelector(
  getSearchState,
  fromSearch.getIds
);
export const getSearchQuery = createSelector(
  getSearchState,
  fromSearch.getQuery
);
export const getSearchLoading = createSelector(
  getSearchState,
  fromSearch.getLoading
);
export const getSearchError = createSelector(
  getSearchState,
  fromSearch.getError
);

export const getSearchResults = createSelector(
  getTaskEntities,
  getSearchTaskIds,
  (tasks, searchIds) => {
    return searchIds.map(id => tasks[id]);
  }
);

export const getCollectionState = createSelector(
  getTasksState,
  (state: TasksState) => state.collection
);

export const getPomosState = createSelector(
  getTasksState,
  (state: TasksState) => state.pomos
);

export const getCollectionLoaded = createSelector(
  getCollectionState,
  fromCollection.getLoaded
);

export const getCollectionLoading = createSelector(
  getCollectionState,
  fromCollection.getLoading
);

export const getCollectionTaskIds = createSelector(
  getCollectionState,
  fromCollection.getIds
);

export const getCollectionPomoIds = createSelector(
  getCollectionState,
  fromCollection.getPomoIds
);

export const getTaskCollection = createSelector(
  getTaskEntities,
  getCollectionTaskIds,
  (entities, ids) => {
    return ids.map(id => entities[id]);
  }
);

export const getPomoCollection = createSelector(
  getPomoEntities,
  getCollectionPomoIds,
  (entities, ids) => {
    return ids.map(id => entities[id]);
  }
);


// export const getTaskPomos = createSelector(
//   getPomoEntities,
//   getCollectionTaskIds,
//   (entities, ids) => {
//     return ids.map(id => entities[id]);
//   }
// );

export const selectTaskPomos = createSelector(
  getAllPomos,
  pomos => (task_id: string) => pomos[task_id]
);



export const getSelectedTaskPomos = createSelector(
  getSelectedTaskId,
  getAllPomos,
  (id, allPomos: Pomo[]) => {
    if (id && getSelectedTaskPomos) {
      return allPomos.filter((pomo) => pomo.task_id.toString() === id);
    } else {
      return allPomos;
    }
  }
);

export const getSelectedTaskPomosTest = createSelector(
  getSelectedTaskId,
  getPomoCollection,
  (id, taskPomos: Pomo[]) => {
    if (taskPomos) {
      return taskPomos.filter(((pomo) => pomo.task_id.toString() === id));
    }});

export const isSelectedTaskInCollection = createSelector(
  getCollectionTaskIds,
  getSelectedTaskId,
  (ids, selected) => {
    ids = ids.map(id => id.toString());
    ids.map(id => id.toString());
    console.log(typeof(ids[0]));
    console.log(typeof(selected));
    console.log([ids]);
    console.log(ids.indexOf(selected));
    return ids.indexOf(selected) > -1;
  }
);


