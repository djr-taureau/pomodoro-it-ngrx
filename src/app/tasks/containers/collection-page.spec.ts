import { CollectionPageComponent } from './collection-page';
import { combineReducers, Store, StoreModule } from '@ngrx/store';
import { async, ComponentFixture, TestBed } from '@angular/core/testing';
import { NoopAnimationsModule } from '@angular/platform-browser/animations';
import { RouterTestingModule } from '@angular/router/testing';
import { MatCardModule, MatInputModule } from '@angular/material';
import { TaskPreviewListComponent } from '../components/Task-preview-list';
import { TaskPreviewComponent } from '../components/Task-preview';
import * as CollectionActions from '../actions/collection';
import * as fromTasks from '../reducers';
import { EllipsisPipe } from '../../shared/pipes/ellipsis';
import { AddCommasPipe } from '../../shared/pipes/add-commas';

describe('Collection Page', () => {
  let fixture: ComponentFixture<CollectionPageComponent>;
  let store: Store<fromTasks.State>;
  let instance: CollectionPageComponent;

  beforeEach(() => {
    TestBed.configureTestingModule({
      imports: [
        NoopAnimationsModule,
        StoreModule.forRoot({
          Tasks: combineReducers(fromTasks.reducers),
        }),
        MatCardModule,
        MatInputModule,
        RouterTestingModule,
      ],
      declarations: [
        CollectionPageComponent,
        TaskPreviewListComponent,
        TaskPreviewComponent,
        TaskAuthorsComponent,
        AddCommasPipe,
        EllipsisPipe,
      ],
    });

    fixture = TestBed.createComponent(CollectionPageComponent);
    instance = fixture.componentInstance;
    store = TestBed.get(Store);

    spyOn(store, 'dispatch').and.callThrough();
  });

  it('should compile', () => {
    fixture.detectChanges();

    expect(fixture).toMatchSnapshot();
  });

  it('should dispatch a collection.Load on init', () => {
    const action = new CollectionActions.Load();

    fixture.detectChanges();

    expect(store.dispatch).toHaveBeenCalledWith(action);
  });
});
