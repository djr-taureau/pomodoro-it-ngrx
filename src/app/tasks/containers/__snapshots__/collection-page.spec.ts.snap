// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Collection Page should compile 1`] = `
<pomo-collection-page
  Tasks$={[Function Store]}
  store={[Function Store]}
>

      <mat-card
    _ngcontent-c0=""
    class="mat-card"
  >


    <mat-card-title
      _ngcontent-c0=""
      class="mat-card-title"
    >
      My Collection
    </mat-card-title>


  </mat-card>

      <pomo-task-preview-list
    _ngcontent-c0=""
    _nghost-c2=""
    ng-reflect-tasks=""
  >





  </pomo-task-preview-list>

</pomo-collection-page>
`;
